src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs

```diff
@@ -12,100 +13,141 @@
 
 namespace Microsoft.CodeAnalysis
 {
+    /// <summary>
+    /// Options used to load <see cref="SourceText"/>.
+    /// </summary>
+    public readonly record struct LoadTextOptions(SourceHashAlgorithm ChecksumAlgorithm);
+
     /// <summary>
     /// A class that represents access to a source text and its version from a storage location.
     /// </summary>
     public abstract class TextLoader
     {
+        private static readonly ConditionalWeakTable<Type, StrongBox<bool>> s_isObsoleteLoadTextAndVersionAsyncOverriden = new();
+
         private const double MaxDelaySecs = 1.0;
         private const int MaxRetries = 5;
         internal static readonly TimeSpan RetryDelay = TimeSpan.FromSeconds(MaxDelaySecs / MaxRetries);
 
         internal virtual string? FilePath => null;
 
+        /// <summary>
+        /// True if <see cref="LoadTextAndVersionAsync(LoadTextOptions, CancellationToken)"/> reloads <see cref="SourceText"/> from its original binary representation (e.g. file on disk).
+        /// </summary>
+        internal virtual bool CanReloadText
+            => false;
+
+        /// <summary>
+        /// Load a text and a version of the document.
+        /// </summary>
+        /// <param name="options">
+        /// Use <see cref="LoadTextOptions.ChecksumAlgorithm"/> when creating <see cref="SourceText"/> from an original binary representation.
+        /// Ignore if the <see cref="SourceText"/> is not created from a binary representation.
+        /// </param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <exception cref="IOException" />
+        /// <exception cref="InvalidDataException"/>
+        /// <exception cref="OperationCanceledException"/>
+        public virtual Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (s_isObsoleteLoadTextAndVersionAsyncOverriden.GetValue(
+                GetType(),
+                _ => new StrongBox<bool>(new Func<Workspace, DocumentId, CancellationToken, Task<TextAndVersion>>(LoadTextAndVersionAsync).Method.DeclaringType != typeof(TextLoader))).Value)
```

Comment:
>new Func<Workspace, DocumentId, CancellationToken, Task<TextAndVersion>>(LoadTextAndVersionAsync) [](http://example.com/codeflow?start=41&length=97)

Consider enclosing object creation in parenthesis in order to improve readability.